### PR TITLE
add fuzzy search to autocompletion

### PR DESCRIPTION
**Pull request by averrin**
But it has no letter highlights

### CODE REVIEW

**Comment by Carpetsmoker** [comment]
Im my humble opinion this should be an option. I, for one, strongly dislike "fuzzy" searching.

**Comment by averrin** [comment]
I know, its very basic and confusing implementation, just for discussion. I agree, option for this is must have.

**Comment by Carpetsmoker** [comment]
Okay ;-)

A related thing I've been wanting to implement by the way is to search only at the start of urls. For example, If I type "git" I get all URL results with the word "git" in them, but 90% of the time, I would like results that *start* with git (like *git*hub.com), and not other results (like `stackoverflow.com/questions/42-how-do-i-foo-in-git` or `someweblog.org/why-mercurial-is-better-than-git`, etc.)

It's just an idea, but perhaps, if you're working on this anyway, you can set the (new) option to accept 3 values: `match` (match anywhere, like it works now), `fuzzy` (your patch), or `start` (as described above).

**Comment by The-Compiler** [comment]
@averrin do you intend to work on this some more in the near future? If not, I'd rather add some notes to #133 and close this (that's why I said "If you'd like to flesh this out a bit more" in #949 :wink:)

**Comment by averrin** [comment]
I think, i can implement match/fuzzy/start autocompletion. Ler it be open for couple of week, ill try to do it.

**Comment by The-Compiler** [comment]
I wonder if using `fuzzywuzzy` is even the right approach - I don't think we want fuzzy matching with a levenshtein distance, do we?

e.g. you wouldn't want `:sel` to match `:set` - that doesn't make any sense...

I think some approach like the projects I linked in #133 would make much more sense - the blog post there even mentions:

> When I first started looking into fuzzy matching in python, I encountered this excellent library called fuzzywuzzy. But the fuzzy matching done by that library is a different kind. It uses levenshtein distance to find the closest matching string from a collection. Which is a great technique for auto-correction against spelling errors but it doesn't produce the desired results for matching long names from partial sub-strings.

**Comment by averrin** [comment]
So, i added "completion:match-type" option with variants:
* contain - just "pattern in data"
* start - data.startswith(pattern), but ignore protocols in url
* fuzzy - very basic implementation, which just check that data contains all pattern chars

Very simple, but works,

File: qutebrowser/completion/models/sortfilter.py [hunk 1]
```diff
@@ -146,10 +148,44 @@ def filterAcceptsRow(self, row, parent):
...
-                    return True
+                else:
+                    matchers = {
+                        'contain': self.containMatcher,
+                        'fuzzy': self.fuzzyMatcher,
+                        'start': self.startMatcher
+                    }
+                    match_type = config.get('completion', 'match-type')
+                    if match_type not in matchers:
+                        match_type = 'contain'
```
Comment (The-Compiler) [inline qutebrowser/completion/models/sortfilter.py]: Why is this here? If the setting type makes sure it's `contain`/`fuzzy`/`start`, I don't see why this check would be needed.

File: qutebrowser/completion/models/sortfilter.py [hunk 2]
```diff
@@ -146,10 +148,44 @@ def filterAcceptsRow(self, row, parent):
                 data = self.srcmodel.data(idx)
                 if not data:
                     continue
-                elif self.pattern.casefold() in data.casefold():
-                    return True
+                else:
+                    matchers = {
+                        'contain': self.containMatcher,
+                        'fuzzy': self.fuzzyMatcher,
+                        'start': self.startMatcher
```
Comment (The-Compiler) [inline qutebrowser/completion/models/sortfilter.py]: Please call those `_match_contains`, `_match_fuzzy` and `_match_start`. Rationales:

* "matcher" makes it sound like they are some kind of object, while it's just a method.
* `mixedCase` is only used for Qt's methods in qutebrowser, `snake_case` is used otherwise.
* Those aren't needed from outside this class, so they can be "private" (starting with `_`)

File: qutebrowser/completion/models/sortfilter.py [hunk 3]
```diff
@@ -146,10 +148,44 @@ def filterAcceptsRow(self, row, parent):
...
+        data = data.casefold()
+        pattern = self.pattern.casefold()
+        return pattern in data
+
+    def startMatcher(self, data):
+        """Matcher for 'start' matching type logic."""
+        protocol = re.compile('^(?:(ht|f)tp(s?)://)?')
+        data = data.casefold()
+        pattern = self.pattern.casefold()
+        if protocol.match(data).lastindex is not None:
```
Comment (The-Compiler) [inline qutebrowser/completion/models/sortfilter.py]: I think you could make this simpler by removing the outermost group (`(?:...)?`)  and using `if protocol.match(data) is not None` - or am I missing something?

File: qutebrowser/config/configdata.py [hunk 4]
```diff
@@ -402,6 +402,10 @@ def data(readonly=False):
              "Whether to shrink the completion to be smaller than the "
              "configured size if there are no scrollbars."),
 
+            ('match-type',
+             SettingValue(typ.String(none_ok=True), 'contain'),
```
Comment (The-Compiler) [inline qutebrowser/config/configdata.py]: Ah, now I see why you checked for invalid values earlier - this should be its own config type which validates and documents the possible values.

Add something like this at the bottom of `config/configtypes.py`:

```python
class MatchType(BaseType):

    """How to match values in the completion."""

    valid_values = ValidValues(('contain', "Match items which contain the "
                                           "entered text."),
                               ('fuzzy', "Match items with a fuzzy matching "
                                         "logic."),
                               ('start', "Only match items starting with the "
                                         "entered text."),
```

And then use `typ.MatchType()`.

**Comment by averrin** [comment]
I hope, everything were fixed

File: qutebrowser/completion/completiondelegate.py [hunk 5]
```diff
@@ -22,9 +22,8 @@
 We use this to be able to highlight parts of the text.
 """
 
-import re
 import html
-
+import re
```
Comment (The-Compiler) [inline qutebrowser/completion/completiondelegate.py]: This seems like an unrelated change

File: qutebrowser/completion/completiondelegate.py [hunk 6]
```diff
@@ -197,9 +196,20 @@ def _get_textdoc(self, index):
...
-                              flags=re.IGNORECASE)
+                if config.get('completion', 'match-type') == 'fuzzy':
+                    text = ''
+                    for char in self._opt.text:
+                        if char in pattern:
+                            text += '<span class="highlight">%s</span>' % char
+                        else:
+                            text += char
+                else:
+                    repl = r'<span class="highlight">\g<0></span>'
```
Comment (The-Compiler) [inline qutebrowser/completion/completiondelegate.py]: Note to self: This should probably be HTML-escaped correctly - but that's not an issue with your PR, as it was already wrong before.

File: qutebrowser/config/configtypes.py [hunk 7]
```diff
@@ -660,7 +660,8 @@ class Font(BaseType):
                 (?P<size>[0-9]+((\.[0-9]+)?[pP][tT]|[pP][xX]))
             )\                         # size/weight/style are space-separated
         )*                             # 0-inf size/weight/style tags
-        (?P<family>[A-Za-z0-9, "-]*)$  # mandatory font family""", re.VERBOSE)
+        (?P<family>[A-Za-z0-9, "-]*)$  # mandatory font family
+    """, re.VERBOSE)
```
Comment (The-Compiler) [inline qutebrowser/config/configtypes.py]: This seems like an unrelated change too.

**Comment by averrin** [comment]
1. really strange. 0_o
2. yeah, ill think about it (in separate PR)
3. It breaks Atom highlighting and i fix it without breaking pep

If u want, i can fix 1 and 3 in additional commit

**Comment by The-Compiler** [comment]
Okay - would be nice if you reverted 1., you can keep 3. if you prefer.

I'll test the PR later today - I wonder if the fuzzy matching really should be per-char (rather than per-word), but I'll brainstorm and play a bit later.

**Comment by averrin** [comment]
I think better variant of fuzzy search - respect chars order in pattern. I will play with it in couple of days

**Comment by The-Compiler** [comment]
Another thing I was thinking of is that you'd split on words (e.g. using `re.split` with `\w`) and then filter words in the right order.

**Comment by averrin** [comment]
So, i want enter "download-remove".

1. do -> download*
2. dore -> download-remove

Inho, typed char order is enough for good matching

**Comment by The-Compiler** [comment]
That'd be the same when matching the start of *words* - as an example with URLs, say you have:

```
http://example.com/this-is-some-news-article
```

Now I want to match that using `thneart` - letter-order matching would also give me `github.com/msiemens/PyGitUp/tree/master/PyGitUp` (and from my history, 4500 other entries) which have nothing to do with what I wanted to match.

**Comment by averrin** [comment]
Maybe u r right, but only if all entries have separators... but something like "jsoneditoronline.com" will be hard to match.

**Comment by averrin** [comment]
So, with some style-related issues (cant switch from js), i added simple cursor for fuzzy matching. 
I think, fuzzy is best variant as fallback match_type, e.g. if you use contain, but start typing something "fuzzy" and contain matching is unsuccessful.

**Comment by The-Compiler** [comment]
I probably won't get to taking a closer look at this this week, but hopefully next week!

Could you write some tests for this? I see you already discovered some corner cases, so I really think they should be tested. Here is a test file to get you started, as there are no tests for `CompletionFilterModel` yet:

`tests/unit/completion/models/test_sortfilter.py`

```python
# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:

# Copyright 2015 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
#
# This file is part of qutebrowser.
#
# qutebrowser is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# qutebrowser is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.

"""Tests for CompletionFilterModel."""

import pytest

from qutebrowser.completion.models import base, sortfilter


@pytest.mark.parametrize('match_type, pattern, data, expected', [
    ('contain', 'foo', 'barfoobar', True),
    ('contain', 'ab', 'aonebtwo', False),
    ('contain', 'x', 'blah', False),

    ('fuzzy', 'foo', 'barfoobar', True),
    ('fuzzy', 'ab', 'aonebtwo', True),
    ('fuzzy', 'x', 'blah', False),
    ('fuzzy', 'ab', 'boneatwo', False),

    ('start', 'foo', 'foobar', True),
    ('start', 'foo', 'barfoobar', False),
    ('start', 'foo', 'barfoo', False),
    ('start', 'ab', 'aonebtwo', False),
    ('start', 'x', 'blah', False),
])
def test_filter_accepts_row(config_stub, match_type, pattern, data, expected):
    config_stub.data = {'completion': {'match-type': match_type}}

    source_model = base.BaseCompletionModel()
    cat = source_model.new_category('test')
    source_model.new_item(cat, data)

    filter_model = sortfilter.CompletionFilterModel(source_model)
    filter_model.set_pattern(pattern)
    assert filter_model.rowCount() == 1  # "test" category
    idx = filter_model.index(0, 0)
    assert idx.isValid()

    if expected:
        assert filter_model.rowCount(idx) == 1
    else:
        assert filter_model.rowCount(idx) == 0
```

**Comment by averrin** [comment]
Some starter set of tests was added.
Maybe some dangerous and extreme strings is good too, but i think its more essential for other components, like completionWidget

**Comment by The-Compiler** [comment]
Do you think this is usable enough to get merged? I just played with `fuzzy` a bit, and personally found it unusable for history - it shows a lot of stuff, but definitely not what wanted...

**Comment by averrin** [comment]
Hm, im rare user of history search, but its ueful for me. If u dont want merge it in this condition, maybe i should add to config something like "force-contain-match-for-history", or implement separated match-type for all search types?

**Comment by The-Compiler** [comment]
I'm really sorry I didn't reply to this earlier, I got kind of lost with a lot of PRs...

I thought about this some more, and I think the fuzzy matching implemented in #1195 is a lot more usable. [I added the `contains` tests you wrote](https://github.com/The-Compiler/qutebrowser/blob/master/tests/unit/completion/test_sortfilter.py), but I'd rather like to keep the fuzzy matching as-is.

Sorry for the disappointment, and thanks for your contribution either way!
